torch/csrc/distributed/c10d/CUDASymmetricMemoryUtils.cpp

```diff
@@ -0,0 +1,223 @@
+#include <sys/socket.h>
+#include <sys/syscall.h>
+#include <sys/un.h>
+#include <unistd.h>
+
+#include <c10/util/error.h>
+
+#if !defined(USE_ROCM) && defined(PYTORCH_C10_DRIVER_API_SUPPORTED)
+#include <c10/cuda/driver_api.h>
+#endif
+
+#include <torch/csrc/distributed/c10d/CUDASymmetricMemoryUtils.hpp>
+#include <torch/csrc/distributed/c10d/Store.hpp>
+#include <torch/csrc/distributed/c10d/cuda/utils.hpp>
+
+namespace c10d::symmetric_memory {
+
+bool device_has_multicast_support(int device_idx) {
+  if (c10::utils::check_env("TORCH_SYMM_MEM_DISABLE_MULTICAST") == true) {
+    return false;
+  }
+  return c10d::cuda::deviceSupportsMulticast(device_idx);
+}
+
+bool allow_overlapping_devices() {
+  return c10::utils::check_env("TORCH_SYMM_MEM_ALLOW_OVERLAPPING_DEVICES") ==
+      true;
+}
+
+// Query environment variable to get the backend used for CUDA Symmetric Memory.
```

Comment:
what's your plan for documenting this to users?
a) how long-lived is it? (iiuc, it's long-lived since we have to support 'Cuda' backend for AMD
b) what are the user-facing tradeoffs?  Why would i choose CUDASymmetricMemory vs NVSHMEMSymmetricMemory in a given scenario?

Reply:
a) The env itself could be longer term due to possibility of "NCCL" backend. The "CUDASymmetricMemory" impl could be shorter-termed but still useful for the reason you mentioned.
b) intra vs inter capability is the main difference from user point of view.

Reply:
(b) is still not clear to me. Does nvshem backend support intra+inter or is it purely intra?

if nvshmem supports the superset of cuda backend _except_ for AMD support, then I would argue we delete the env, and autodetect AMD to switch it to the cuda backend. otherwise, default to the nvshmem backend?

Reply:
The env is also for NCCLSymmetricMemory vs NVSHMEMSymmetricMemory.